tools/clang/tools/libclang/dxcrewriteunused.cpp

```diff
@@ -301,6 +305,10 @@ class TypeVisitor : public RecursiveASTVisitor<TypeVisitor> {
     }
     return true;
   }
+  bool VisitArrayType(const ArrayType *AT) {
+    TraverseType(AT->getElementType());
+    return true;
+  }
```

Comment:
The corresponding DataRecursiveASTVistor code is a bit hard to follow due to macro skullduggery, but I don't think this is ever called. The only use of TypeVisitor is https://github.com/microsoft/DirectXShaderCompiler/blob/c2233c6c0a528ecaf5d2d008684e08daf722d845/tools/clang/tools/libclang/dxcrewriteunused.cpp#L1462
The implementation of which is here
https://github.com/microsoft/DirectXShaderCompiler/blob/c2233c6c0a528ecaf5d2d008684e08daf722d845/tools/clang/include/clang/AST/DataRecursiveASTVisitor.h#L548
Which defines ABSTRACT_TYPE as nothing and for Arrays, this is the macro that Array uses:
https://github.com/microsoft/DirectXShaderCompiler/blob/c2233c6c0a528ecaf5d2d008684e08daf722d845/tools/clang/include/clang/AST/TypeNodes.def#L70

So I really don't see how this could be called. Am I missing something?

Reply:
VisitArrayType will be called here https://github.com/microsoft/DirectXShaderCompiler/blob/c2233c6c0a528ecaf5d2d008684e08daf722d845/tools/clang/include/clang/AST/DataRecursiveASTVisitor.h#L350

But the change is still useless because VisitRecordType will be invoked anyway with/without the change. :)

Removed the VisitArrayType.